File: data_bridge_s3/components/data_bridge/test_apps/test_data_bridge.py
from pytest_embedded import Dut
import struct 


# def test_debug(dut: Dut):
#     print(type(dut.serial.proc))
#     print(dir(dut.serial.proc))

def test_unity(dut: Dut):
    dut.expect_exact("PASS")

    dut.expect_exact("READY_SERIAL")
    packet = struct.pack(
    "<BB??H",
    1, # node_id
    2, # sensor_id
    1, # trigger_shutter
    0, # measure_distance
    1234 # distance_mm
    )
    # dut.write(packet)
    dut.write(packet)
    dut.expect_exact(
        "node=1 sensor=2 shutter=1 measure=0 distance=1234"
    )
    

File: data_bridge_s3/components/data_bridge/test_apps/main/test_data_bridge.c
#include "unity.h"
#include "data_packet.h"
#include "driver/usb_serial_jtag.h"
#include "freertos/FreeRTOS.h"
#include "freertos/task.h"

#define BUF_SIZE (1024)

static void test_packet_size(void)
{
    TEST_ASSERT_EQUAL(
        6,
        sizeof(data_packet_t));
}


void test_serial_parsing(void)
{
    usb_serial_jtag_driver_config_t cfg = {
    .rx_buffer_size = BUF_SIZE,
    .tx_buffer_size = BUF_SIZE,
    };

    ESP_ERROR_CHECK(
        usb_serial_jtag_driver_install(&cfg));

    printf("READY_SERIAL");
    printf("\n");

    data_packet_t packet;

    int len = usb_serial_jtag_read_bytes(
        (uint8_t *)&packet,
        sizeof(packet),
        pdMS_TO_TICKS(10000));

    printf("Length: %u \n", len);

    if (len == sizeof(data_packet_t))
    {
        printf(
        "node=%u sensor=%u shutter=%u measure=%u distance=%u\n",
        packet.node_id,
        packet.sensor_id,
        packet.trigger_shutter,
        packet.measure_distance,
        packet.distance_mm);
    }
}

void app_main(void)
{
    UNITY_BEGIN();
    RUN_TEST(test_packet_size);
    RUN_TEST(test_serial_parsing);

    UNITY_END();
}